### PR TITLE
External extensions

### PR DESCRIPTION
**Pull request by claytongulick**
## Description

Add support for simplified external extensions, external build and debugging and vanilla Web Components for extension UI.

Fixes #

## Type of Change

- [ ] Bugfix
- [ X] Improvement
- [ ] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [ ] New / updated tests are included
- [ X] All tests are passing locally
- [X ] Performed a self-review of the submitted code

If adding a new feature:

- [ ] Documentation was added/updated

### CODE REVIEW

**Comment by claytongulick** [comment]
Hi @rijkvanzanten here's a clean PR, sorry about the clutter in the last one.

So far, this enhancement has been working great for us.

File: app/src/components/v-form/form-field-interface.vue [hunk 1]
```diff
@@ -24,6 +24,8 @@
 			:collection="field.collection"
 			:field="field.field"
 			:field-data="field"
+			:values="values"
```
Comment (u12206050) [inline app/src/components/v-form/form-field-interface.vue:27]: `values` is already provided and can be injected in sub components of the form using `inject('values')` 
Can you rather use that.

File: app/src/components/v-form/form-field-interface.vue [hunk 2]
```diff
@@ -77,14 +80,30 @@ export default defineComponent({
 			type: Boolean,
 			default: false,
 		},
+		values: {
```
Comment (u12206050) [inline app/src/components/v-form/form-field-interface.vue:83]: See comment above. To get values just do `const values = inject('values')`

File: app/src/components/v-form/form-field-interface.vue [hunk 3]
```diff
@@ -77,14 +80,30 @@ export default defineComponent({
...
 
 		const interfaceExists = computed(() => !!getInterface(props.field?.meta?.interface || 'input'));
 
-		return { t, interfaceExists, getDefaultInterfaceForType };
+		const api = useApi();
+		const { useCollectionsStore, useFieldsStore } = useStores();
+		const collectionsStore = useCollectionsStore();
+		const fieldsStore = useFieldsStore();
+
+		let context = {
```
Comment (u12206050) [inline app/src/components/v-form/form-field-interface.vue:99]: Creating a context object here isn't really saving anyone work, instead it goes against the provide and inject pattern.
Each sub field/display can access stores and the api and emit themselves IF they need it, they don't need to start sending around a context ;)

File: app/src/components/v-form/form-field.vue [hunk 4]
```diff
@@ -38,6 +38,7 @@
 			:batch-mode="batchMode"
 			:batch-active="batchActive"
 			:disabled="isDisabled"
+			:values="values"
```
Comment (u12206050) [inline app/src/components/v-form/form-field.vue:41]: Remove and use `inject`

File: app/src/components/v-form/form-field.vue [hunk 5]
```diff
@@ -104,6 +106,7 @@ const props = withDefaults(defineProps<Props>(), {
 	validationError: undefined,
 	autofocus: false,
 	badge: undefined,
+	values: {},
```
Comment (u12206050) [inline app/src/components/v-form/form-field.vue:109]: Remove and use `inject`

File: app/src/components/v-form/form-field.vue [hunk 6]
```diff
@@ -91,6 +92,7 @@ interface Props {
 	validationError?: ValidationError;
 	autofocus?: boolean;
 	badge?: string;
+	values: any;
```
Comment (u12206050) [inline app/src/components/v-form/form-field.vue:95]: Remove and use `inject`

File: app/src/components/v-form/v-form.vue [hunk 7]
```diff
@@ -54,6 +54,7 @@
 				:batch-active="batchActiveFields.includes(field.field)"
 				:primary-key="primaryKey"
 				:loading="loading"
+				:values="values"
```
Comment (u12206050) [inline app/src/components/v-form/v-form.vue:57]: Remove and use `inject`

Comment (u12206050) [reply app/src/components/v-form/v-form.vue:57]: Actually this you can just remove. v-form is actually the component that provides the values object

**Review by u12206050 — CHANGES_REQUESTED** [review]
Could you give a use case & context for using this. What do you mean by external extensions?

**Comment by claytongulick** [comment]
> Could you give a use case & context for using this. What do you mean by external extensions?

Sure thing, this comes from a discussion with @rijkvanzanten and several others about the need to be able to have a single, unopioninated repository that multiple extensions can be loaded from.

The current solution duplicates dependencies because each component is build independently and then combined later in a rollup call into a massive js file that's delivered. As a simple example, I have multiple extensions that use the form.io library, it's big. I can't have that dep duplicated, I had a js file that was 15 megs!

Also, Vue works great with web components. Since that's the case, there's no reason to enforce extension authors using vue. This PR also makes some enhancements to support the web components use case. 

Sorry this wasn't all clear, this was a cleaned up PR and lost some context of the original conversation.

File: app/src/components/v-form/form-field-interface.vue (hunk 1 above)
Comment (claytongulick) [reply app/src/components/v-form/form-field-interface.vue:27]: But we don't all use vue 😄 

Assigning values like that allows a setting on a vanilla web component to participate in the fun.

File: app/src/components/v-form/form-field-interface.vue (hunk 3 above)
Comment (claytongulick) [reply app/src/components/v-form/form-field-interface.vue:99]: Again, that's Vue-only. For the vanilla use case, injection doesn't work.

**Comment by u12206050** [comment]
> > Could you give a use case & context for using this. What do you mean by external extensions?
> 
> @u12206050 Sure thing, this comes from a discussion with @rijkvanzanten and several others about the need to be able to have a single, unopioninated repository that multiple extensions can be loaded from.
> 
> The current solution duplicates dependencies because each component is build independently and then combined later in a rollup call into a massive js file that's delivered. As a simple example, I have multiple extensions that use the form.io library, it's big. I can't have that dep duplicated, prior to implementing this, I had a js file that was 15 megs!
> 
> Also, Vue works great with web components. Since that's the case, there's no reason to enforce extension authors using vue. This PR also makes some enhancements to support the web components use case, this allows the use of any framework that compiles down to WCs (most of them these days), and any build system (webpack, etc...) as long as the interface for the extension metadata is honored in the default export.
> 
> Sorry this wasn't all clear, this was a cleaned up PR and lost some context of the original conversation.

Thanks for explaining. I like the idea, although I feel trying to mix just any web component directly with Directus without wrapping it could lead to some difficult to solve use cases.

Couldn't this better be solved with the Directus-extension wrapping your custom component? That wrapper would then inject all required variables such as stores, api, and values and pass that on. You would anyways need a wrapper to define the component id, name, icon and other options etc.

By wrapping each component as an extension we shield Directus from the affects of having to always "include" variables for non-vue components both now and in the future (ie. when new features or refactoring is being done.)

In regards to optimising the form.io library you could load that library from a cdn and have all your components use that instead of bundling it with your code. Here is an article that might help: https://www.mixmax.com/engineering/rollup-externals

Sorry if I am bring up points that you already discussed with @rijkvanzanten

**Comment by rijkvanzanten** [comment]
Heya, thanks for your PR! Our small team has been attempting to work through a backlog of external contributions for the better part of a year, but in that time many have become too stale to merge or the change is no longer desired/relevant.. 
We've updated the [contributing guidelines](https://github.com/directus/directus/blob/main/contributing.md#pull-requests) and are working on [making more resources available](https://github.com/directus/directus/discussions/17977) to make sure that doesn’t happen again 😬 In the meantime, we'll have to close this PR for now..

**Pull request closed without merging**